client/core/trade.go

```diff
@@ -735,6 +797,63 @@ func (t *trackedTrade) redeemMatches(matches []*matchTracker) error {
 	return errs.ifany()
 }
 
+// refundMatches will send refund transactions for the specified matches.
+func (t *trackedTrade) refundMatches(matches []*matchTracker) (uint64, error) {
+	errs := newErrorSet("refundMatches: order %s - ", t.ID())
+
+	refundWallet, refundAsset := t.wallets.fromWallet, t.wallets.fromAsset // refunding to our wallet
+	var refundedQty uint64
+
+	for _, match := range matches {
+		dbMatch, metaData, proof, _ := match.parts()
+		contractToRefund := proof.Script
+		var swapCoinID order.CoinID
+		var matchFailureReason string
+		switch {
+		case dbMatch.Side == order.Maker && metaData.Status == order.MakerSwapCast:
+			swapCoinID = proof.MakerSwap
+			matchFailureReason = "no valid counterswap received from Taker"
+		case dbMatch.Side == order.Taker && metaData.Status == order.TakerSwapCast:
+			swapCoinID = proof.TakerSwap
+			matchFailureReason = "no valid redemption received from Maker"
+		default:
+			log.Errorf("attempted to execute invalid refund for match %s", match.id)
+			continue
+		}
+
+		swapCoinString := coinIDString(refundAsset.ID, swapCoinID)
+		log.Infof("failed match, %s, refunding %s contract %s",
+			matchFailureReason, unbip(refundAsset.ID), swapCoinString)
+
+		err := refundWallet.Refund(dex.Bytes(swapCoinID), contractToRefund, refundAsset)
+		if err != nil {
+			if err == asset.CoinSpentError {
+				// todo: begin find redemption
+				// note: may be that swap was actually refunded, so FindRedemption
+				// should account for that.
+			}
+			errs.add("error sending refund tx for match %s, swap coin %s: %v",
+				match.id, swapCoinString, err)
+			match.failErr = err
+			continue
+		}
+		if t.Trade().Sell {
+			refundedQty += match.Match.Quantity
+		} else {
+			refundedQty += calc.BaseToQuote(match.Match.Rate, match.Match.Quantity)
+		}
+		// todo: order.MatchFailed? Status needs to change so that refund is
+		// not re-attempted on trackedTrade.tick().
```

Comment:
Not seeing it yet... is there an advantage to keeping the `{Actor}SwapCast` status after successful refund? Historical reasons perhaps?

Reply:
Is there any case in which a taker could go to a refund from any status other than `TakerSwapCast`, or a maker from an status other than `MakerSwapCast`? Now that I think about it, I don't think that there is. But we do want to save the refund coin for historical purposes.

Reply:
I see we're not returning the coin ID from `Refund` yet.

Reply:
I've added that.

Reply:
I think you have it working as is now, but I do think we should do something. I wouldn't want to use `MatchComplete` though, as I would consider that an indicator of success. I think it's okay in this case to add a new `MatchStatus` (since unlike a revocation, the status progression is unambiguous) if that works for everyone else.

Reply:
Yeah, saving the refund coin prevents re-attempting to refund; while keeping the status may help in later review of where the trade went south. Maybe no need to change the status?

> Is there any case in which a taker could go to a refund from any status other than `TakerSwapCast`, or a maker from an status other than `MakerSwapCast`?

Missed this earlier. Maker could also refund if status is `TakerSwapCast` but maker was unable to redeem the Taker's swap for some reason. I see this isn't yet considered in this PR.